File: jkmvc-common/src/main/kotlin/net/jkcode/jkmvc/common/_BitSet.java
package net.jkcode.jkmvc.common

import java.util.*

/**
 *
 * @author shijianhang<772910474@qq.com>
 * @date 2019-06-26 7:18 PM
 */
class BitIndexIterator(protected val bits:BitSet): Iterator<Int> {

    protected var pos = 0

    override fun hasNext(): Boolean {
        bits.nextSetBit(pos)
    }

    override fun next(): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}

File: jkmvc-common/src/main/kotlin/net/jkcode/jkmvc/common/_BitSet.kt
package net.jkcode.jkmvc.common

import net.jkcode.jkmvc.bit.SetBitIterator
import java.util.*

/**
 * 获得比特集的迭代器
 * @return
 */
public fun BitSet.iterator(): SetBitIterator {
    return SetBitIterator(this)
}

/**
 * Performs the given [operation] on each element of this [SetBitIterator].
 * @sample samples.collections.Iterators.forEachIterator
 */
public inline fun BitSet.forEach(action: (Int) -> Unit): Unit {
    iterator().forEach(action)
}

/**
 * Returns a list containing the results of applying the given [transform] function
 * to each element in the original collection.
 */
public inline fun <R> BitSet.map(transform: (Int) -> R): List<R> {
    return iterator().map(transform)
}